hack/verify-upgrade.sh

```diff
@@ -0,0 +1,188 @@
+#!/usr/bin/env bash
+
+# Copyright 2021 The cert-manager Authors.
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#     http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+set -o nounset
+set -o errexit
+set -o pipefail
+
+SCRIPT_ROOT="$( cd "$( dirname "${BASH_SOURCE[0]}" )" > /dev/null && pwd )"
+export REPO_ROOT="${SCRIPT_ROOT}/.."
+source "${REPO_ROOT}/devel/lib/lib.sh"
+source "${REPO_ROOT}/devel/lib/cert.sh"
+source "${REPO_ROOT}/hack/build/version.sh"
+
+kube::version::get_version_vars
+
+LATEST_RELEASE="${KUBE_LAST_RELEASE}"
+CURRENT_VERSION="${KUBE_GIT_VERSION}"
+
+# Ensure helm, kind, kubectl, ytt, jq are available
+bazel build //hack/bin:helm //hack/bin:kind //hack/bin:ytt //hack/bin:jq //hack/bin:kubectl
+bindir="$(bazel info bazel-bin)"
+export PATH="${bindir}/hack/bin/:$PATH"
+
+echo "Testing upgrade from ${LATEST_RELEASE} to ${CURRENT_VERSION}"
+
+# Namespace to deploy into
+NAMESPACE="${NAMESPACE:-cert-manager}"
+# Release name to use with Helm
+RELEASE_NAME="${RELEASE_NAME:-cert-manager}"
+# cert-manager Helm chart
+HELM_CHART="jetstack/cert-manager"
+
+############
+# VERIFY INSTALL, UPGRADE, UNINSTALL WITH HELM
+############
+
+# This will target the host's helm repository cache
+helm repo add jetstack https://charts.jetstack.io
+helm repo update
+
+# 1. INSTALL THE LATEST PUBLISHED HELM CHART
+
+echo "Installing cert-manager ${LATEST_RELEASE} Helm chart into the cluster..."
+
+# Upgrade or install latest published cert-manager Helm release
+helm upgrade \
+    --install \
+    --wait \
+    --namespace "${NAMESPACE}" \
+    --set installCRDs=true \
+    --create-namespace \
+    --version "${LATEST_RELEASE}" \
+    "$RELEASE_NAME" \
+    "$HELM_CHART"
+
+# TODO: This is a dumb sleep that must be removed once we have a check for webhook to be ready.
+echo "Sleep for 30s whilst waiting for cert-manager installation to be ready..."
+sleep 30
+
+# Create a cert-manager issuer and cert
+kubectl apply -f "${REPO_ROOT}/test/fixtures/cert-manager-resources.yaml" \
+	--selector=test=first
+
+# Ensure cert becomes ready
+ensure_cert_ready "test1"
+
+# 1. BUILD AND UPGRADE TO HELM CHART FROM THE CURRENT MASTER
+
+echo "Upgrading cert-manager Helm release to ${CURRENT_VERSION}..."
+"${REPO_ROOT}/devel/addon/certmanager/install.sh"
```

Comment:
I can't remember whether this waits until the new version of the webhook is running and has acquired the leader election resource.

I guess it must since you haven't added the sleep 30 here.

Reply:
It runs `helm upgrade` [with `--wait` flag](https://github.com/jetstack/cert-manager/blob/master/devel/addon/certmanager/install.sh#L58) which would wait for the webhook deployment to be ready, I think that there is no leader election for webhook.
TBH, I don't understand why we seem to only error if applying CM resources right after an initial CM deploy, but never after a rolled out upgrade.

Reply:
.. but I have anyway changed it to wait for apply to succeed after an upgrade too, just to avoid potential flakes.